File: day7/day7.py
#!/usr/bin/env python3

import re
import sys

pat = re.compile(r"([a-z]+)|\[([a-z]+)\]")

def is_abba(sub):
    for i in range(len(sub)-3):
        if sub[i] == sub[i+3] and sub[i+1] == sub[i+2] \
           and sub[i] != sub[i+1]:
            return True

    return False

def supports_tls(sequence):
    found = False
    for a, b in pat.findall(sequence):
        if a and is_abba(a):
            found = True
        elif b and is_abba(b):
            return False

    return found

def find_aba(sub):
    lst = []
    for i in range(len(sub)-2):
        if sub[i] != sub[i+1] and sub[i] == sub[i+2]:
            lst.append(sub[i:i+3])
    return lst

def supports_ssl(sequence):
    aba = []
    bab = []
    for a, b in pat.findall(sequence):
        if a:
            aba.extend(find_aba(a))
        else:
            for b in find_aba(b):
                bab.append(b[1]+b[0]+b[1])

    for a in aba:
        if a in bab:
            return True
    return False

if __name__ == "__main__":
    if len(sys.argv) < 2:
        print("Usage: {} <filename>".format(sys.argv[0]), file=sys.stderr)
        sys.exit(1)

    try:
        with open(sys.argv[1], "rt") as f:
            txt = f.read().strip()
    except:
        print("Cannot open {}".format(sys.argv[1]), file=sys.stderr)
        sys.exit(1)

    tls = 0
    ssl = 0
    for row in txt.splitlines():
        if supports_tls(row):
            tls += 1
        if supports_ssl(row):
            ssl += 1

    print("Part1:", tls)
    print("Part2:", ssl)

File: day7/day7.cpp
#include <fstream>
#include <vector>

#include <fmt/format.h>

using namespace std;

static bool is_abba(const string& str, size_t begin, size_t end)
{
	if (end > str.size())
	{
		end = str.size();
	}
	for (size_t i = begin+3; i < end; i++)
	{
		if (str[i-3] == str[i]
		    && str[i-2] == str[i-1]
		    && str[i] != str[i-1])
		{
			return true;
		}
	}
	return false;
}

static bool supports_tls(const string& ip)
{
	size_t pos = 0;
	size_t next;
	enum {OUT, IN} state = OUT;
	bool ret = false;
	do
	{
		if (state == OUT)
		{
			next = ip.find('[', pos);
			ret = ret || is_abba(ip, pos, next);
			state = IN;
		}
		else
		{
			next = ip.find(']', pos);
			if (is_abba(ip, pos, next))
			{
				return false;
			}
			state = OUT;
		}
		pos = next + 1;
	} while (next != string::npos);

	return ret;
}

static size_t count_tls(const vector<string>& ips)
{
	return count_if(
		ips.begin(),
		ips.end(),
		[](const auto& ip) {
			return supports_tls(ip);
		});
}

static void find_aba(const string& str, size_t begin, size_t end, vector<pair<char,char>>& aba)
{
	if (end > str.size())
	{
		end = str.size();
	}
	for (size_t i = begin+2; i < end; i++)
	{
		if (str[i] == str[i-2] && str[i] != str[i-1])
		{
			aba.emplace_back(make_pair(str[i], str[i-1]));
		}
	}
}

static bool supports_ssl(const string& ip)
{
	size_t pos = 0;
	size_t next;
	enum {OUT, IN} state = OUT;
	vector<pair<char,char>> aba;
	vector<pair<char,char>> bab;
	do
	{
		if (state == OUT)
		{
			next = ip.find('[', pos);
			find_aba(ip, pos, next, aba);
			state = IN;
		}
		else
		{
			next = ip.find(']', pos);
			find_aba(ip, pos, next, bab);
			state = OUT;
		}
		pos = next + 1;
	} while (next != string::npos);

	for (auto [a1, a2]: aba)
	{
		for(auto [b1, b2]: bab)
		{
			if (a1 == b2 && a2 == b1)
			{
				return true;
			}
		}
	}
	return false;
}

static size_t count_ssl(const vector<string>& ips)
{
	return count_if(
		ips.begin(),
		ips.end(),
		[](const auto& ip) {
			return supports_ssl(ip);
		});
}

int main(int argc, char *argv[])
{
	if (argc < 2)
	{
		fmt::print(stderr, "Usage: {} <filename>\n", argv[0]);
		return 1;
	}

	ifstream input(argv[1]);
	if (!input)
	{
		fmt::print(stderr, "Cannot open {}\n", argv[1]);
		return 1;
	}
	vector<string> ips;
	string ip;
	while (getline(input, ip))
	{
		ips.emplace_back(move(ip));
	}
	input.close();

	fmt::print("Part1: {}\nPart2: {}\n", count_tls(ips), count_ssl(ips));
	return 0;
}
